File: src/favoringIsolatedMutability/java/filesize/ConcurrentFileSizeWAkka.java
/***
 * Excerpted from "Programming Concurrency on the JVM",
 * published by The Pragmatic Bookshelf.
 * Copyrights apply to this code. It may not be used to create training material, 
 * courses, books, articles, and the like. Contact us if you are in doubt.
 * We make no guarantees that this code is fit for any purpose. 
 * Visit http://www.pragmaticprogrammer.com/titles/vspcon for more book information.
***/
package com.agiledeveloper.pcj;

import akka.actor.UntypedActor;
import akka.actor.UntypedActorFactory;
import akka.actor.ActorRef;
import akka.actor.Actors;
import akka.actor.ActorRegistry;
import java.util.List;
import java.util.ArrayList;
import java.io.File;

class RequestAFile {}

class FileSize {
  public final long size;
  public FileSize(final long fileSize) { size = fileSize; }
}

class FileToProcess {
  public final String fileName;
  public FileToProcess(final String name) { fileName = name; }
}

class FileProcessor extends UntypedActor {
  private final ActorRef sizeCollector;
  
  public FileProcessor(final ActorRef theSizeCollector) { 
    sizeCollector = theSizeCollector; 
  }
  
  @Override public void preStart() { registerToGetFile(); }
  
  public void registerToGetFile() { 
    sizeCollector.sendOneWay(new RequestAFile(), getContext()); 
  }
  
  public void onReceive(final Object message) {
    FileToProcess fileToProcess = (FileToProcess) message;
    final File file = new java.io.File(fileToProcess.fileName);
    long size = 0L;
    if(file.isFile()) {
      size = file.length();
    } else {
      File[] children = file.listFiles();
      if (children != null)
        for(File child : children)
          if (child.isFile())
            size += child.length();
          else
            sizeCollector.sendOneWay(new FileToProcess(child.getPath()));
    }
      
    sizeCollector.sendOneWay(new FileSize(size));
    registerToGetFile();
  }
}

class SizeCollector extends UntypedActor {
  private final List<String> toProcessFileNames = new ArrayList<String>();
  private final List<ActorRef> idleFileProcessors = 
    new ArrayList<ActorRef>();
  private long pendingNumberOfFilesToVisit = 0L;
  private long totalSize = 0L;
  private long start = System.nanoTime();

  public void sendAFileToProcess() {
    if(!toProcessFileNames.isEmpty() && !idleFileProcessors.isEmpty())
      idleFileProcessors.remove(0).sendOneWay(
        new FileToProcess(toProcessFileNames.remove(0)));
  } 
   
  public void onReceive(final Object message) {
    if (message instanceof RequestAFile) {
      idleFileProcessors.add(getContext().getSender().get());
      sendAFileToProcess();
    }
      
    if (message instanceof FileToProcess) {
      toProcessFileNames.add(((FileToProcess)(message)).fileName);
      pendingNumberOfFilesToVisit += 1;
      sendAFileToProcess();
    }

    if (message instanceof FileSize) {
      totalSize += ((FileSize)(message)).size;
      pendingNumberOfFilesToVisit -= 1;
      
      if(pendingNumberOfFilesToVisit == 0) {        
        long end = System.nanoTime();
        System.out.println("Total size is " + totalSize);
        System.out.println("Time taken is " + (end - start)/1.0e9);    
        Actors.registry().shutdownAll();    
      }
    }
  }
}

public class ConcurrentFileSizeWAkka {
  public static void main(final String[] args) { 
    final ActorRef sizeCollector = 
      Actors.actorOf(SizeCollector.class).start();
      
    sizeCollector.sendOneWay(new FileToProcess(args[0]));
    
    for(int i = 0; i < 100; i++) 
      Actors.actorOf(new UntypedActorFactory() {
        public UntypedActor create() {
          return new FileProcessor(sizeCollector); 
        }
      }).start();        
  }
}

File: bin/concurrency-code/favoringIsolatedMutability/scala/filesize/ConcurrentFileSizeWAkka.scala
package com.agiledeveloper.pcj

import akka.actor.Actor
import akka.actor.ActorRef
import akka.actor.Actors
import akka.actor.ActorRegistry

//START:CASECLASSES
case object RequestAFile
case class FileSize(size : Long)
case class FileToProcess(fileName : String)
//END:CASECLASSES

//START:FILEPROCESSOR
class FileProcessor(val sizeCollector : ActorRef) extends Actor {
  override def preStart = registerToGetFile
  
  def registerToGetFile = { sizeCollector ! RequestAFile }
  
  def receive = {
    case FileToProcess(fileName) =>
      val file = new java.io.File(fileName)

      var size = 0L
      if(file.isFile()) {
        size = file.length()
      } else {
        val children = file.listFiles()
        if (children != null)
          for(child <- children)
            if (child.isFile())
              size += child.length()
            else
              sizeCollector ! FileToProcess(child.getPath())
      }
      
      sizeCollector ! FileSize(size)
      registerToGetFile    
  }
}
//END:FILEPROCESSOR

//START:SIZECOLLECTOR
class SizeCollector extends Actor {
  var toProcessFileNames = List.empty[String]
  var fileProcessors = List.empty[ActorRef]
  var pendingNumberOfFilesToVisit = 0L
  var totalSize = 0L
  val start = System.nanoTime()

  def sendAFileToProcess() : Unit = {
    if(!toProcessFileNames.isEmpty && !fileProcessors.isEmpty) {
      fileProcessors.head ! FileToProcess(toProcessFileNames.head)
      fileProcessors = fileProcessors.tail
      toProcessFileNames = toProcessFileNames.tail      
    }
  }    
  def receive = {
    case RequestAFile =>
      fileProcessors = self.getSender().get :: fileProcessors
      sendAFileToProcess()     
    case FileToProcess(fileName) =>
      toProcessFileNames = fileName :: toProcessFileNames
      pendingNumberOfFilesToVisit += 1
      sendAFileToProcess()
    case FileSize(size) =>
      totalSize += size
      pendingNumberOfFilesToVisit -= 1
      if(pendingNumberOfFilesToVisit == 0) {
        val end = System.nanoTime()
        println("Total size is " + totalSize)
        println("Time taken is " + (end - start)/1.0e9)        
        Actors.registry.shutdownAll
      }
  }
}
//END:SIZECOLLECTOR

//START:MAIN
object ConcurrentFileSizeWAkka {
  def main(args : Array[String]) : Unit = {
    val sizeCollector = Actor.actorOf[SizeCollector].start()
    
    sizeCollector ! FileToProcess(args(0))
    
    for(i <- 1 to 100) 
      Actor.actorOf(new FileProcessor(sizeCollector)).start()
  }  
}
//END:MAIN
